### ISSUE TITLE
Add Installation-HowTo for Raspberry PI

### ISSUE THREAD

**Issue by kmuehlbauer** [vispy/vispy-website#16]
Add HowTo describing the installation process for at least one backend to get VisPy running on the Raspberry Pi.

**Comment by 0xF0F0F0F0** [comment]
Hi,

I stumbled upon this while trying to make VisPy run on the Pi 3 model B. I managed to run the cube example without installing through the berryconda env, but i'm getting around 6-7 FPS. It isn't using the GPU, CPU utilization was at 98%. 

I then tried installing pyopengl and pyqt though berryconda, but was unable to find the pyqt package. I am not familiar with building from recipes so I am stuck here. 

Would you be able to provide further steps from here? We could then perhaps document it and add to the "How To".

**Comment by kmuehlbauer** [comment]
@0xF0F0F0F0 Please read also the discussion in https://github.com/vispy/vispy/issues/495. I did not have the time to further investigate :cry:.

**Comment by 0xF0F0F0F0** [comment]
I did go through that discussion, and saw that you managed to successfully run vispy on the pi3 model B through berryconda. 

I was able to follow through and install everything in the conda env except for pyqt. 

How did you build pyqt in the conda environment?

**Comment by kmuehlbauer** [comment]
@0xF0F0F0F0  Have a look [here](https://github.com/jjhelmus/berryconda/issues/15) too. I think I used the pyqt-recipe from berryconda to build pyqt. I did not manage to build `qt` but used the system qt libraries instead, IIRC.

**Comment by kmuehlbauer** [comment]
@djhoese Where do I should add installation instructions for raspberry? Add a dedicated `Raspberry Pi` section under http://vispy.org/installation.html#?

**Comment by djhoese** [comment]
Sounds good to me.

**Comment by kmuehlbauer** [comment]
OK, this will take a little time, as I want to reproduce every step with a clean install with latest updates etc.

In the end it will look something like this:

Raspberry Pi 3
==============

This section covers installing VisPy on a Raspberry Pi 3 Model B using [Rasbian Stretch with Desktop](https://www.raspberrypi.org/downloads/raspbian/). Download the image and write it on a fast micro SDHC (class 10) with enough capacity (at least 16GB).

Prerequisites
-------------

- set to full KMS driver in `raspi-config`
- clone [berryconda repository]((https://github.com/jjhelmus/berryconda))
- download and install [berryconda3](https://github.com/jjhelmus/berryconda/releases/download/v2.0.0/Berryconda3-2.0.0-Linux-armv7l.sh)
- install necessary `Qt4`-packages and other needed packages via raspbian package manager (also devel packages)
- install mesa-tools

Check OpenGL capabilities
-------------------------

- check opengl capabilities using `glxinfo` and `glxgears` (and possibly more)

Build PyOpenGL and PyQt4
------------------------

- PyOpenGl should be available with berryconda, but building it here should also be OK
- use the `pyopengl`-recipe to conda-build `PyOpenGL`-package
- use the `pyqt4`-recipe to conda-build `PyQt4`

Install Vispy
-------------

- clone vispy-repository
- create a dedicated `vispy`-environment with conda
- conda install pyopengl, pyqt4 and numpy
- install vispy

**Comment by djhoese** [comment]
Maybe since this looks like it is going to be long it would be good to add it as a separate page and link to it from the installation page. Also does berryconda have the ability to use other channels so you could put your PyQt4 build/recipe on a public channel?

**Comment by kmuehlbauer** [comment]
@djhoese OK separate page. And yes, I plan to do so, if we can't get it into berryconda main channel. Need to check if and how this works.

**Comment by djhoese** [comment]
This is making me want to get a raspberry pi 😜

**Comment by Coimbra1984** [comment]
Hi,

I'm coming back to this issue. I was reading this issue: https://github.com/vispy/vispy/issues/495 and that one: https://github.com/jjhelmus/berryconda/issues/15
I managed to get vispy to work in a berryconda environemnt with pyqt4.
I think that I followed your steps mentioned above. What I did was:
1. enable full KMS in raspi-config
2. download berryconda
3. install qt4 host libraries
4. create a vispy environment
5. install pyopengl with pip
6. build pyqt4 with custom recipies
7. pip install vispy
8. modify and run https://github.com/vispy/vispy/blob/master/examples/basics/gloo/rotate_cube.py (I added c.measure_fps())

But I only get ~10FPS and have >300% of CPU load on my PI:
10.2 FPS
11.3 FPS
11.9 FPS
12.3 FPS
12.6 FPS
13.3 FPS
12.5 FPS
12.2 FPS
12.1 FPS
11.9 FPS
11.9 FPS

Do you have any hints on what I'm doing wrong?

Thanks for the great work so far!

**Comment by kmuehlbauer** [comment]
@lemarquois Yes, it seems you are running on CPU. Maybe it's No. 5 on the list. AFAIR I installed pyopengl via conda/conda-build not pip.
I did not had the time yet to redo the whole process with a fresh SDCard, but this on my plate.

**Comment by kmuehlbauer** [comment]
@lemarquois I might have made changes to the conda recipes. This might be another source of problems. Anyway I need to buy a new SDCard to not overwrite my working version.

**Comment by Coimbra1984** [comment]
okay, I have found the root cause on my system.
The problem was not related to vispy at all.
Running glxgears also gave very poor performance (expected value is ~60fps but I had something around 20fps). This seems to be a common problem on raspbian stretch distribution (the former distribution raspbian jessie doesn't have this problem) .
In the end, I read a lot of forum posts on the raspberry tryied out a lof of recompiles. I even built mesa on my own, but the bug mentioned here: https://bugs.launchpad.net/raspbian/+bug/1713746 solved my problem:
A simple
`sudo mv /usr/share/X11/xorg.xonf.d/99-fbturbo.conf /usr/share/X11/xorg.xonf.d/99-fbturbo.conf.bak`

An important check seems to be `glxinfo` as well. The line `OpenGL renderer string` should not contain llvmpipe  this seems to be some kind of software rendering (I'm not familiar with openGL).

In the end, I think you can even use the pyqt4 and pyopengl standard libraries from raspberry, in my case it was really the problem of getting glxgears to perform as expected.

**Comment by kmuehlbauer** [comment]
@lemarquois @djhoese 

Finally, I have acquired a new SDHC. Following @lemarquois last remark, i tried to get things running with raspbian stretch standard libraries . And it works for most of the examples. My procedure so far can be found [here](https://gist.github.com/kmuehlbauer/91703349fd84067304bc40b1d540c1df).

One thing was different, I did not have to move this file. It just worked. But that might be due to the fact, that the raspberry had full KMS enabled before first starting the raspbian system.

I will try to check other backends as well (without using X11).
